internal/pkg/manifest/svc.go

```diff
@@ -524,55 +556,104 @@ func (cfg NetworkLoadBalancerListener) exposedPorts(exposedPorts []ExposedPort,
 	targetProtocol = strings.ToLower(targetProtocol)
 	for _, exposedPort := range exposedPorts {
 		if targetPort == exposedPort.Port && targetProtocol == exposedPort.Protocol {
-			return nil, nil
+			return nil
 		}
 	}
 	targetContainer := workloadName
 	if cfg.TargetContainer != nil {
 		targetContainer = aws.StringValue(cfg.TargetContainer)
 	}
 
-	return []ExposedPort{
-		{
-			Port:          targetPort,
-			Protocol:      targetProtocol,
-			ContainerName: targetContainer,
-		},
-	}, nil
+	exposedPorts[targetPort] = ExposedPort{
+		Port:          targetPort,
+		Protocol:      targetProtocol,
+		ContainerName: targetContainer,
+	}
+
+	return nil
 }
 
-func (sidecar SidecarConfig) exposedPorts(sidecarName string) ([]ExposedPort, error) {
+func (sidecar SidecarConfig) exposePorts(exposedPorts map[uint16]ExposedPort, sidecarName string) error {
 	if sidecar.Port == nil {
-		return nil, nil
+		return nil
 	}
 	sidecarPort, protocolPtr, err := ParsePortMapping(sidecar.Port)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	protocol := aws.StringValue(protocolPtr)
 	if protocolPtr == nil {
 		protocol = "tcp"
 	}
 	port, err := strconv.ParseUint(aws.StringValue(sidecarPort), 10, 16)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return []ExposedPort{
-		{
-			Port:                 uint16(port),
-			Protocol:             strings.ToLower(protocol),
-			ContainerName:        sidecarName,
-			isDefinedByContainer: true,
-		},
-	}, nil
+
+	exposedPorts[uint16(port)] = ExposedPort{
+		Port:                 uint16(port),
+		Protocol:             strings.ToLower(protocol),
+		ContainerName:        sidecarName,
+		isDefinedByContainer: true,
+	}
+
+	return nil
 }
 
-func sortExposedPorts(exposedPorts []ExposedPort) []ExposedPort {
-	// Sort the exposed ports so that the order is consistent and the integration test won't be flaky.
-	sort.Slice(exposedPorts, func(i, j int) bool {
-		return exposedPorts[i].Port < exposedPorts[j].Port
-	})
-	return exposedPorts
+// ServiceConnectTargetContainer contains the name of a container and port to expose to ECS Service Connect.
+type ServiceConnectTargetContainer struct {
+	Container string
+	Port      string
+}
+
+// ServiceConnectTarget returns the target container, port, and protocol to be exposed for ServiceConnect.
+func (l *LoadBalancedWebService) ServiceConnectTarget(exposedPorts ExposedPortsIndex) *ServiceConnectTargetContainer {
+	// Expose ServiceConnect from `image.port` by default.
+	targetContainer := exposedPorts.WorkloadName
+	targetPort := exposedPorts.mainContainerPort()
+	targetProtocol := exposedPorts.mainContainerProtocol()
+
```

Comment:
I wonder if it makes sense to return early if `l.HTTPOrBool.Disabled()`!

I agree that whether we add the check of not, the result of the execution would be the same. However, I feel like it aligns with our mental model better: 1. We use `image.port` as the sc port by default; 2. If there is a known port that's designated to handle non-udp traffic, then we use that port for sc.

The issue with today's `l.HTTPOrBool.Main.exposedContainerAndPort` is that, this function always assumes that HTTP is enabled when it's called: it'd still happily return a port, even if we have `http: false`. This issue happens for both LBWS and Backend service.

Therefore, to make sure that our logic is sound, the caller needs to make sure that the assumption is true - that when we call this function, indeed, HTTP is enabled. 

Let me know if this gibberish above doesn't make sense to you 😂 Basically, adding the check is just for the sake of the logic.